### PR TITLE
*: clone AggDesc before modifying its `Mode` in AggDesc.Split

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
Before this commit, when enabling prepare plan cache, the following sql will panic:
``` sql
create table t(a int, b int);
insert into t value(1,1),(2,2);
prepare stmt from 'select distinct a from t order by a';

execute stmt;
execute stmt; -- This will fail.
```

### What is changed and how it works?

Why PlanCache will impact the behavior of the above sql?
1. If the aggregate function can be executed parallelly, AggFuncDesc.Split will be called when building HashAggExec to split the AggFuncDesc into partialAggDesc and finalAggDesc.
2. Before AggFuncDesc.Split, the AggFuncMode will **ONLY** be CompleteMode and FinalMode.
3. After AggFuncDesc.Split, the origin AggFuncDesc.Mode will be changed into Partial1Mode or Partial2Mode.
4. PreparePlanCache would cache the physical plan, when we use the cached plan to build the HashAggExec again, the AggFuncDesc.Mode has been modified in `step 3`.

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch
release-2.1

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -1075,7 +1075,7 @@ func (b *executorBuilder) buildHashAgg(v *plannercore.PhysicalHashAgg) Executor
 				ordinal = append(ordinal, partialOrdinal+1)
 				partialOrdinal++
 			}
-			finalDesc := aggDesc.Split(ordinal)
+			finalDesc := aggDesc.Split(ordinal, sessionVars.StmtCtx.UseCache)
```
Comment (eurekaka) [inline executor/builder.go]: Could you explain the root cause a little bit more? Why would `UseCache` impact the behavior of executor?

**Comment by winoros** [comment]
Is this caused by that the `AggFuncDesc` modified its `Mode` variable?
What about making `Split`'s return value as `(finalAggDesc, partialAggDesc)` to avoid to modify itself?

**Comment by shenli** [comment]
@XuHuaiyu Do not forget to cherry-pick to the release-2.1 branch.

**Comment by XuHuaiyu** [comment]
@shenli ok

**Pull request merged**